code/workspaces/infrastructure-api/src/controllers/projectsController.js

```diff
@@ -0,0 +1,108 @@
+import { check, matchedData } from 'express-validator';
+import { service } from 'common';
+import projectsRepository from '../dataaccess/projectsRepository';
+
+async function listProjects(request, response, next) {
+  try {
+    const projects = await projectsRepository.getAll();
+    response.send(projects);
+  } catch (error) {
+    next(new Error(`Error listing projects: ${error.message}`));
+  }
+}
+
+async function getProjectByKey(request, response, next) {
+  const { projectKey } = matchedData(request);
+
+  try {
+    const project = await projectsRepository.getByKey(projectKey);
+    if (project) {
+      response.send(project);
+    } else {
+      response.status(404).send();
+    }
+  } catch (error) {
+    next(new Error(`Error getting project by key: ${error.message}`));
+  }
+}
+
+async function createProject(request, response, next) {
+  const project = matchedData(request, { locations: ['body'] });
+
+  if (await projectsRepository.exists(project.projectKey)) {
+    response.status(400).send({
+      errors: [{ msg: `Entry with projectKey '${project.projectKey}' already exists.` }],
+    });
+  } else {
+    try {
+      const createdProject = await projectsRepository.create(project);
+      response.status(201).send(createdProject);
+    } catch (error) {
+      next(new Error(`Error creating project: ${error.message}`));
+    }
+  }
+}
+
+async function createOrUpdateProject(request, response, next) {
+  const { projectKey } = matchedData(request, { locations: ['params'] });
```

Comment:
I thought that `matchedData` with no location by default looked in all locations?

Reply:
You're right it does, but there is a `projectKey` in the body of the request as well so want to explicitly get the one in the params.